File: nodejs/src/Tree/Diameter/Diameter.ts
import BaseTreeAlgorithmT from "../BaseTreeAlgorithm";
import TreeNodeT from "../TreeNode";

class Diameter<T> extends BaseTreeAlgorithmT<T>{
  exec(): number{
    // final result
    let res: number = 0;

    // The depth of a node in the tree is the length of the path from the root to the node.
    const depth = (node: TreeNodeT<T> | null): number => {
      // base case
      if(node === null) return 0;

      // find left & right depth
      const left: number = depth(node.left)
      const right: number = depth(node.right)

      // update result with left + right
      res = Math.max(left + right, res)

      // update depth with max value
      return Math.max(left, right) + 1
    }

    depth(this.tree.root)

    return res
  }
}

export default Diameter;

File: go/src/tree/diameter/diameter.go
package diameter

import "algorithms/go/tree"

var res int

func depth(node *tree.Node) int {
	if node == nil {
		return 0
	}

	// get left and right edge count
	var left int = depth(node.Left)
	var right int = depth(node.Right)
	// update res
	res = max(left+right, res)

	// recursive formula for depth
	return max(left, right) + 1
}

func diameterOfBinaryTree(root *tree.Node) int {
	// note the update here
	res = 0

	depth(root)

	return res
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}
